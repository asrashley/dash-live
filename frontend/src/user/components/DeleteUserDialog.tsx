import { useContext } from "preact/hooks";
import { AppStateContext } from "../../appState";
import { ModalDialog } from "../../components/ModalDialog";

export interface DeleteUserDialogProps {
  username: string;
  onCancel: () => void;
  onConfirm: () => void;
}

export function DeleteUserDialog({
  username,
  onCancel,
  onConfirm,
}: DeleteUserDialogProps) {
  const { dialog } = useContext(AppStateContext);

  if (!dialog.value?.confirmDelete) {
    return null;
  }
  const footer = (
    <div className="form-actions">
      <button className="btn btn-danger me-4" data-testid="confirm-delete" onClick={onConfirm}>
        Delete {username}
      </button>
      <button className="btn btn-primary me-2" data-testid="cancel-delete" onClick={onCancel}>
        Cancel
      </button>
    </div>
  );

  return (
    <ModalDialog
      onClose={onCancel}
      title="Are you sure you want to delete this user?"
      footer={footer}
    >
      <h3>Delete user "{username}" ?</h3>
    </ModalDialog>
  );
}
