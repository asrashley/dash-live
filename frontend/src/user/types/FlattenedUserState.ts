import type { InitialUserState } from "../types/InitialUserState";


export type FlattenedUserState = Omit<InitialUserState, 'groups'> & {
  adminGroup: boolean;
  mediaGroup: boolean;
  userGroup: boolean;
};
