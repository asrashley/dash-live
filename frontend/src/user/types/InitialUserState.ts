export type InitialUserState = {
    pk?: number;
    username?: string;
    email?: string;
    lastLogin: string | null;
    mustChange: boolean;
    groups: string[];
};
