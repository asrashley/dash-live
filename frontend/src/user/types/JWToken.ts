export type JWToken = {
  expires: string;
  jwt: string;
};
