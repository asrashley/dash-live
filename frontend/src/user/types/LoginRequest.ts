export interface LoginRequest {
    username: string;
    password: string;
    rememberme: boolean;
}