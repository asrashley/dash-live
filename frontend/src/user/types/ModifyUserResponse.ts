import type { InitialUserState } from "./InitialUserState";

export interface ModifyUserResponse {
    success: boolean;
    errors: string[];
    user?: InitialUserState;
}
