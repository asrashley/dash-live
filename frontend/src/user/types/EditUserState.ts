import { FlattenedUserState } from "./FlattenedUserState";


export type EditUserState = FlattenedUserState & {
  password?: string;
  confirmPassword?: string;
};
