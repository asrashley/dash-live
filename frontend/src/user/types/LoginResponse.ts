import { JWToken } from "./JWToken";
import { InitialUserState } from "./InitialUserState";

export interface LoginResponse {
    success: boolean;
    error?: string;
    csrf_token: string;
    accessToken?: JWToken;
    refreshToken?: JWToken;
    user?: InitialUserState
}
