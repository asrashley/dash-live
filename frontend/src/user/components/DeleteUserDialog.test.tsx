import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { computed, signal } from "@preact/signals";
import userEvent from "@testing-library/user-event";

import { renderWithProviders } from "../../test/renderWithProviders";
import { AppStateType } from "../../appState";
import { DialogState } from "../../types/DialogState";

import { DeleteUserDialog } from "./DeleteUserDialog";

describe("DeleteUserDialog component", () => {
  const onCancel = vi.fn();
  const onConfirm = vi.fn();
  const closeDialog = vi.fn();
  const dialog = signal<DialogState | null>(null);
  const cinemaMode = signal<boolean>(false);
  const appState: AppStateType = {
    cinemaMode,
    dialog,
    backdrop: computed(() => !!dialog.value?.backdrop),
    closeDialog,
  };
  const username = "new.user";

  beforeEach(() => {
    dialog.value = {
      backdrop: true,
      confirmDelete: {
        name: username,
        confirmed: false,
      },
    };
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("shows dialog", () => {
    const { getBySelector, asFragment } = renderWithProviders(
      <DeleteUserDialog
        username={username}
        onCancel={onCancel}
        onConfirm={onConfirm}
      />,
      { appState }
    );
    getBySelector(".modal-dialog");
    expect(asFragment()).toMatchSnapshot();
  });

  test("hides dialog", () => {
    dialog.value = {
      backdrop: false,
    };
    const { queryBySelector } = renderWithProviders(
      <DeleteUserDialog
        username={username}
        onCancel={onCancel}
        onConfirm={onConfirm}
      />,
      { appState }
    );
    expect(queryBySelector(".modal-dialog")).toBeNull();
  });

  test("can confirm dialog", async () => {
    const user = userEvent.setup();
    const { getByTestId } = renderWithProviders(
      <DeleteUserDialog
        username={username}
        onCancel={onCancel}
        onConfirm={onConfirm}
      />,
      { appState }
    );
    const btn = getByTestId("confirm-delete") as HTMLButtonElement;
    await user.click(btn);
    expect(onCancel).not.toHaveBeenCalled();
    expect(onConfirm).toHaveBeenCalledTimes(1);
  });

  test("can cancel dialog", async () => {
    const user = userEvent.setup();
    const { getByTestId } = renderWithProviders(
      <DeleteUserDialog
        username={username}
        onCancel={onCancel}
        onConfirm={onConfirm}
      />,
      { appState }
    );
    const btn = getByTestId("cancel-delete") as HTMLButtonElement;
    await user.click(btn);
    expect(onCancel).toHaveBeenCalledTimes(1);
    expect(onConfirm).not.toHaveBeenCalled();
  });
});
