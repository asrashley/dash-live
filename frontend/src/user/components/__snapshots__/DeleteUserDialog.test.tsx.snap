// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DeleteUserDialog component > shows dialog 1`] = `
<DocumentFragment>
  <div
    aria-modal="true"
    class="modal fade show"
    id="dialog-box"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
      role="document"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
            Are you sure you want to delete this user?
          </h5>
          <button
            aria-label="Close"
            class="close btn-close"
            type="button"
          />
        </div>
        <div
          class="modal-body m-2"
        >
          <h3>
            Delete user "new.user" ?
          </h3>
        </div>
        <div
          class="modal-footer"
        >
          <div
            class="form-actions"
          >
            <button
              class="btn btn-danger me-4"
              data-testid="confirm-delete"
            >
              Delete new.user
            </button>
            <button
              class="btn btn-primary me-2"
              data-testid="cancel-delete"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
